### ISSUE TITLE
OrmLite: second bug in Sql.In (new issue instead of reopen)

### ISSUE THREAD

**Issue by OlegNadymov** [ServiceStack/Issues#454]
This issue with tests. It is for https://github.com/ServiceStack/Issues/issues/446

Three falied tests:

```C#
    public class BaseObject
    {
        public Guid? Identity { get; set; }
    }

    public class Waybill : BaseObject
    {
        private readonly IDbConnection _db = new OrmLiteConnection(new OrmLiteConnectionFactory("test", SqlServerDialect.Provider));

        public string Name { get; set; }

        public void TestMethod1()
        {
            var localIdentity = this.Identity;
            var expr = _db.From<MarginItem>().Where(s => Sql.In(s.Identity,
                _db.From<WaybillItem>()
                    .Where(b => b.WaybillId == localIdentity)
                    .Select(b => b.MarginItemId)));

            expr.ToSelectStatement().Dump();
            
            //Problem 1 (the variable name is inserted in the SQL, instead of a parameter):
            /*
             SELECT "MarginName", "Identity" 
            FROM "MarginItem"
            WHERE "Identity" IN (SELECT "MarginItemId" 
            FROM "WaybillItem"
            WHERE ("WaybillId" = "localIdentity"))
            */
        }

        public void TestMethod2()
        {
            var expr = _db.From<MarginItem>().Where(s => Sql.In(s.Identity,
                _db.From<WaybillItem>()
                    .Where(b => b.WaybillId == this.Identity)
                    .Select(b => b.MarginItemId)));

            expr.ToSelectStatement().Dump();

            //Problem 2 (the property name is inserted in the SQL, instead of a parameter):
            /*
             SELECT "MarginName", "Identity" 
            FROM "MarginItem"
            WHERE "Identity" IN (SELECT "MarginItemId" 
            FROM "WaybillItem"
            WHERE ("WaybillId" = "Identity"))
            */
        }

        public void TestMethod3()
        {
            var expr = _db.From<MarginItem>().Where(s => Sql.In(s.Identity,
                _db.From<WaybillItem>()
                    .LeftJoin<MarginItem>((wi, mi) => wi.MarginItemId == mi.Identity)
                    .Where(b => b.WaybillId == this.Identity)
                    .And<MarginItem>(b => b.MarginName == this.Name)
                    .Select(b => b.MarginItemId)));

            expr.ToSelectStatement().Dump();

            //Problem 3 (the "Object". + property names is inserted in the SQL, instead of a parameter):
            /*
             SELECT "MarginName", "Identity" 
            FROM "MarginItem"
            WHERE "Identity" IN (SELECT "WaybillItem"."MarginItemId" 
            FROM "WaybillItem" 
            LEFT JOIN "MarginItem" ON ("WaybillItem"."MarginItemId" = "MarginItem"."Identity")
            WHERE ("WaybillItem"."WaybillId" = "Object"."Identity") AND ("MarginItem"."MarginName" = "Object"."Name"))
            */
        }
    }

    public class WaybillItem : BaseObject
    {
        public string WbItemName { get; set; }

        [ForeignKey(typeof(Waybill))]
        public Guid WaybillId { get; set; }

        [ForeignKey(typeof(MarginItem))]
        public Guid? MarginItemId { get; set; }
    }

    public class MarginItem : BaseObject
    {
        public string MarginName { get; set; }
    }


    //-------------------------
    public class Person2
    {
        [AutoIncrement]
        public int Id { get; set; }

        public string Name { get; set; }
    }

    public class Order2
    {
        [AutoIncrement]
        public int Id { get; set; }

        [References(typeof(Person2))]
        public int Person2Id { get; set; }

        public DateTime Order2Date { get; set; }

        public int Order2TypeId { get; set; }
    }


    public class MyTests
    {
        [Test]
        public void FailedSqlInTest()
        {
            var w = new Waybill()
            {
                Identity = Guid.Empty,
                Name = "WaybillTest"
            };

            w.TestMethod1(); //failed

            w.TestMethod2(); //failed

            w.TestMethod3(); //failed
        }

        [Test]
        public void CorrertSqlInTest()
        {
            int orderTypeId = 2;
            var db = new OrmLiteConnection(new OrmLiteConnectionFactory("test", SqlServerDialect.Provider));

            var subExpr = db.From<Order2>()
                .Where(y => y.Order2TypeId == orderTypeId)
                .Select(y => y.Person2Id);

            //Correct
            subExpr.ToSelectStatement().Dump();
            /*
            SELECT "PersonId" 
            FROM "Order"
            WHERE ("OrderTypeId" = @0)
            */

            var expr = db.From<Person2>()
                .Where(x => Sql.In(x.Id, subExpr));

            //Solved
            expr.ToSelectStatement().Dump();
            /*
            SELECT "Id", "Name" 
            FROM "Person"
            WHERE "Id" IN (SELECT "PersonId" 
            FROM "Order"
            WHERE ("OrderTypeId" = "@0"))
            */
        }
    }
```

And sometimes raises exception with the message: 
```C#
"Unhandled expression type: 'Index'"
```
But I don't understand when it happens.

**Comment by xplicit** [comment]
As workaround you can create subexpression first and then pass it to `Sql.In` statement similar to your `CorrectSqlInTest` method. Here is the sample for your TestMethod1 code:

            var localIdentity = this.Identity;
            var subExpr = _db.From<WaybillItem>()
                    .Where(b => b.WaybillId == localIdentity)
                    .Select(b => b.MarginItemId);

            var expr = _db.From<MarginItem>().Where(s => Sql.In(s.Identity, subExpr));

It generates this SQL code:

    SELECT "MarginName", "Identity" 
    FROM "MarginItem"
    WHERE "Identity" IN (SELECT "MarginItemId" 
    FROM "WaybillItem"
    WHERE ("WaybillId" = @0))

**Comment by xplicit** [comment]
This is regression issue. First occurence of it exists in this commit https://github.com/ServiceStack/ServiceStack.OrmLite/commit/7331ec117034de03f471a4f8be9f113a7839487e

**Comment by mythz** [comment]
This should now be resolved with this commit: https://github.com/ServiceStack/ServiceStack.OrmLite/commit/ffe988cf39a4d5c5d553cced6e3638af8b020d3e

this change is available from v4.0.61 that's now [available on MyGet](https://github.com/ServiceStack/ServiceStack/wiki/MyGet).

**Comment by OlegNadymov** [comment]
Now it works correctly. Thanks!
